envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/ServiceTagsAndCanaryTest.kt

```diff
@@ -1,47 +1,61 @@
-import org.junit.jupiter.api.BeforeAll
 import org.junit.jupiter.api.Test
-import pl.allegro.tech.servicemesh.envoycontrol.config.EnvoyControlRunnerTestApp
-import pl.allegro.tech.servicemesh.envoycontrol.config.EnvoyControlTestConfiguration
+import org.junit.jupiter.api.extension.RegisterExtension
+import pl.allegro.tech.servicemesh.envoycontrol.assertions.isOk
+import pl.allegro.tech.servicemesh.envoycontrol.assertions.untilAsserted
+import pl.allegro.tech.servicemesh.envoycontrol.config.EnvoyControlExtension
+import pl.allegro.tech.servicemesh.envoycontrol.config.consul.ConsulExtension
 import pl.allegro.tech.servicemesh.envoycontrol.config.containers.EchoContainer
+import pl.allegro.tech.servicemesh.envoycontrol.config.containers.EchoServiceExtension
+import pl.allegro.tech.servicemesh.envoycontrol.config.envoy.CallStats
+import pl.allegro.tech.servicemesh.envoycontrol.config.envoy.EnvoyExtension
 
-open class ServiceTagsAndCanaryTest : EnvoyControlTestConfiguration() {
-
+class ServiceTagsAndCanaryTest {
     companion object {
-        private val properties = mapOf(
-            "envoy-control.envoy.snapshot.routing.service-tags.enabled" to true,
-            "envoy-control.envoy.snapshot.routing.service-tags.metadata-key" to "tag",
-            "envoy-control.envoy.snapshot.load-balancing.canary.enabled" to true,
-            "envoy-control.envoy.snapshot.load-balancing.canary.metadata-key" to "canary",
-            "envoy-control.envoy.snapshot.load-balancing.canary.metadata-value" to "1"
-        )
-
-        @JvmStatic
-        @BeforeAll
-        fun setupTest() {
-            setup(appFactoryForEc1 = { consulPort ->
-                EnvoyControlRunnerTestApp(properties = properties, consulPort = consulPort)
-            })
-        }
-
-        private val loremRegularContainer = echoContainer
-        private val loremCanaryContainer = echoContainer2
-        private val ipsumRegularContainer = EchoContainer().also { it.start() }
 
-        @JvmStatic
-        @AfterAll
-        fun cleanup() {
-            ipsumRegularContainer.stop()
-        }
+        @JvmField
+        @RegisterExtension
+        val consul = ConsulExtension()
+
+        @JvmField
+        @RegisterExtension
+        val envoyControl = EnvoyControlExtension(consul, mapOf(
+                "envoy-control.envoy.snapshot.routing.service-tags.enabled" to true,
+                "envoy-control.envoy.snapshot.routing.service-tags.metadata-key" to "tag",
+                "envoy-control.envoy.snapshot.load-balancing.canary.enabled" to true,
+                "envoy-control.envoy.snapshot.load-balancing.canary.metadata-key" to "canary",
+                "envoy-control.envoy.snapshot.load-balancing.canary.metadata-value" to "1"
+        ))
+
+        @JvmField
+        @RegisterExtension
+        val envoy = EnvoyExtension(envoyControl)
+
+        @JvmField
+        @RegisterExtension
+        val loremRegularService = EchoServiceExtension()
```

Comment:
So we will stop using properly shared containers `echoContainer1` and `echoContainer2` from BaseEnvoyTest, instead creating EchoContainer every time we start each test class. I wonder what effect it will have on the tests time when we migrate all the tests to the new approach and will start creating these containers like 40 times instead of just one. You can't really see what impact of the tests time it will have until you migrate all of them to the new approach.

There is no doubt we should use extensions for "fake shared" resources, like Envoy containers, which have shared field but are recreated anyway for every test class (this type of resources are mainly located in EnvoyTestConfiguration class and have `lateinit var` modifier)
It is another story for properly shared resources like echo containers (they are mainly located in BaseEnvoyTest class and have `val` modifier). 

Maybe we should check what resources are really used by majority of the tests and if they are stateless, like EchoContainer, we can keep sharing them.

Reply:
That's a valid concern. I will address it later, as described in "Future work":
> Sharing containers between test classes when possible to speed up the execution.

Reply:
sure, I missed the description, my bad